File: android/src/main/java/io/linkrunner/LinkrunnerModule.java
package io.linkrunner;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;

import io.linkrunner.models.CapturePaymentRequest;
import io.linkrunner.models.RemovePaymentRequest;
import io.linkrunner.models.UserDataRequest;
import io.linkrunner.sdk.LinkRunner;
import io.linkrunner.sdk.models.response.AttributionData;
import io.linkrunner.sdk.models.IntegrationData;
import io.linkrunner.utils.MapUtils;
import io.linkrunner.utils.ModelConverter;

import java.lang.reflect.Method;
import java.util.Map;
import kotlin.Result;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;

import androidx.annotation.NonNull;

public class LinkrunnerModule extends ReactContextBaseJavaModule {
    private static final String TAG = "LinkrunnerModule";

    private static final String MODULE_NAME = "LinkrunnerSDK";
    private final ReactApplicationContext reactContext;
    private final LinkRunner linkrunnerSDK;

    public LinkrunnerModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        // Initialize the native SDK instance
        this.linkrunnerSDK = LinkRunner.getInstance();
    }

    @Override
    public String getName() {
        return MODULE_NAME;
    }

    @ReactMethod
    public void init(String token, ReadableMap options, Promise promise) {

        try {
            String link = null;
            String source = null;
            String secretKey = null;
            String keyId = null;

            // Extract optional parameters if provided
            if (options != null) {
                if (options.hasKey("link")) {
                    link = options.getString("link");
                }

                if (options.hasKey("source")) {
                    source = options.getString("source");
                }

                if (options.hasKey("secretKey")) {
                    secretKey = options.getString("secretKey");
                }

                if (options.hasKey("keyId")) {
                    keyId = options.getString("keyId");
                }
            }

            if (token == null || token.isEmpty()) {
                promise.reject("INIT_ERROR", "Token is required");
                return;
            }

            try {
                Log.i(TAG, "init: Initializing LinkRunner SDK");
                
                linkrunnerSDK.initFromJava(
                        reactContext,
                        token,
                        link,
                        source,
                        secretKey,
                        keyId,
                        baseResponse -> {
                            try {
                                // Create a response map for React Native
                                WritableMap response = Arguments.createMap();
                                
                                // Add status and message
                                response.putString("status", "success");
                                response.putString("message", "Linkrunner SDK initialized successfully");
                                promise.resolve(response);
                                Log.i(TAG, "init: SDK initialized successfully");
                                
                                return Unit.INSTANCE;
                            } catch (Exception e) {
                                Log.e(TAG, "init: Error in success callback: " + e.getMessage(), e);
                                promise.reject("CALLBACK_ERROR", "Error in success callback: " + e.getMessage(), e);
                                return Unit.INSTANCE;
                            }
                        },
                        exception -> {
                            promise.reject("INIT_ERROR", "Failed to initialize Linkrunner: " + exception.getMessage(), exception);
                            return Unit.INSTANCE;
                        }
                );
                Log.i(TAG, "init: initFromJava call completed (async operation started)");
            } catch (Exception e) {
                Log.e(TAG, "init: Exception calling initFromJava: " + e.getMessage(), e);
                promise.reject("INIT_ERROR", "Exception calling initFromJava: " + e.getMessage(), e);
            }
        } catch (Exception e) {
            promise.reject("INIT_ERROR", "Failed to initialize Linkrunner: " + e.getMessage(), e);
        }
    }

    @ReactMethod
    public void signup(ReadableMap userData, ReadableMap data, Promise promise) {
        try {
            // Convert ReadableMap to our model
            UserDataRequest userDataRequest = ModelConverter.toUserDataRequest(userData);

            // Convert ReadableMap to native Map for additionalData
            Map<String, Object> additionalData = null;
            if (data != null && !data.toHashMap().isEmpty()) {
                additionalData = MapUtils.readableMapToMap(data);
            }

            // Call native SDK signup method with proper callback handling
            linkrunnerSDK.signupFromJava(
                userDataRequest.toKotlinModel(),
                additionalData,
                baseResponse -> {
                    Log.i(TAG, "signup: Success callback received");
                    try {
                        // Create response map
                        WritableMap response = Arguments.createMap();
                        
                        response.putString("status", "success");
                        response.putString("message", "User signed up successfully");
                        
                        if (baseResponse != null) {
                            if (baseResponse.getMessage() != null) {
                                response.putString("message", baseResponse.getMessage());
                            }
                        }

                        promise.resolve(response);
                        return Unit.INSTANCE;
                    } catch (Exception e) {
                        promise.reject("CALLBACK_ERROR", "Error in signup success callback: " + e.getMessage(), e);
                        return Unit.INSTANCE;
                    }
                },
                exception -> {
                    Log.e(TAG, "signup: Error callback received: " + exception.getMessage(), exception);
                    promise.reject("SIGNUP_ERROR", "Failed to signup user: " + exception.getMessage(), exception);
                    return Unit.INSTANCE;
                }
            );
        } catch (Exception e) {
            promise.reject("SIGNUP_ERROR", "Failed to signup user: " + e.getMessage(), e);
        }
    }

    @ReactMethod
    public void setUserData(ReadableMap userData, Promise promise) {
        try {
            // Convert ReadableMap to our model
            UserDataRequest userDataRequest = ModelConverter.toUserDataRequest(userData);

            // Call native SDK setUserData method with proper callback
            linkrunnerSDK.setUserDataFromJava(
                userDataRequest.toKotlinModel(),
              () -> {

                    WritableMap response = Arguments.createMap();
                    response.putString("status", "success");
                    response.putString("message", "User data set successfully");

                    // Resolve the promise with the properly converted response
                    promise.resolve(response);
                    return Unit.INSTANCE;
                },
                exception -> {
                    promise.reject("SET_USER_DATA_ERROR", "Failed to set user data: " + exception.getMessage(), exception);
                    return exception;
                }
            );
        } catch (Exception e) {
            promise.reject("SET_USER_DATA_ERROR", "Failed to set user data: " + e.getMessage(), e);
        }
    }

    @ReactMethod
    public void trackEvent(String eventName, ReadableMap eventData, Promise promise) {
        try {
            Map<String, Object> eventDataMap = null;
            if (eventData != null) {
                eventDataMap = MapUtils.readableMapToMap(eventData);
            }

            // Call native SDK trackEvent method with proper callback
            linkrunnerSDK.trackEventFromJava(
                eventName,
                eventDataMap,
              () -> {
                    // Convert the Kotlin GeneralResponse to a JavaScript-friendly WritableMap
//                    WritableMap response = io.linkrunner.utils.ModelConverter.fromGeneralResponse();

                    // Add status and message for better client-side handling
                    WritableMap response = Arguments.createMap();
                    response.putString("status", "success");
                    response.putString("message", "Event tracked successfully");

                    // Resolve the promise with the properly converted response
                    promise.resolve(response);
                    return Unit.INSTANCE;
                },
                exception -> {
                    promise.reject("TRACK_EVENT_ERROR", "Failed to track event: " + exception.getMessage(), exception);
                    return Unit.INSTANCE;
                }
            );
        } catch (Exception e) {
            promise.reject("TRACK_EVENT_ERROR", "Failed to track event: " + e.getMessage(), e);
        }
    }

    @ReactMethod
    public void capturePayment(ReadableMap paymentData, Promise promise) {
        try {
            // Convert ReadableMap to our model
            CapturePaymentRequest capturePaymentRequest = ModelConverter.toCapturePaymentRequest(paymentData);

            // Call native SDK capturePayment method with proper callback
            linkrunnerSDK.capturePaymentFromJava(
                capturePaymentRequest.toKotlinModel(),
              () -> {
                    // Convert the Kotlin GeneralResponse to a JavaScript-friendly WritableMap
                    WritableMap response = Arguments.createMap();


                    response.putString("status", "success");
                    response.putString("message", "Payment captured successfully");

                    // Resolve the promise with the properly converted response
                    promise.resolve(response);
                    return Unit.INSTANCE;
                },
                exception -> {
                    promise.reject("CAPTURE_PAYMENT_ERROR", "Failed to capture payment: " + exception.getMessage(), exception);
                    return Unit.INSTANCE;
                }
            );
        } catch (Exception e) {
            promise.reject("CAPTURE_PAYMENT_ERROR", "Failed to capture payment: " + e.getMessage(), e);
        }
    }

    @ReactMethod
    public void removePayment(ReadableMap paymentData, Promise promise) {
        try {
            // Convert ReadableMap to our model
            RemovePaymentRequest removePaymentRequest = ModelConverter.toRemovePaymentRequest(paymentData);

            // Call native SDK removePayment method with proper callback
            linkrunnerSDK.removePaymentFromJava(
                removePaymentRequest.toKotlinModel(),
              () -> {
                    // Convert the Kotlin GeneralResponse to a JavaScript-friendly WritableMap
                    WritableMap response = Arguments.createMap();
                    response.putString("status", "success");
                    response.putString("message", "Payment removed successfully");
                    promise.resolve(response);
                    return Unit.INSTANCE;
                },
                exception -> {
                    promise.reject("REMOVE_PAYMENT_ERROR", "Failed to remove payment: " + exception.getMessage(), exception);
                    return Unit.INSTANCE;
                }
            );
        } catch (Exception e) {
            promise.reject("REMOVE_PAYMENT_ERROR", "Failed to remove payment: " + e.getMessage(), e);
        }
    }
    
    @ReactMethod
    public void getAttributionData(Promise promise) {
        try {
            // Call native SDK getAttributionDataFromJava method with proper callback
            linkrunnerSDK.getAttributionDataFromJava(
                attributionData -> {
                    try {
                        // Convert the attribution data to a WritableMap
                        WritableMap response = Arguments.createMap();
                        
                        if (attributionData != null) {
                            // Add the deeplink if it exists
                            if (attributionData.getDeeplink() != null) {
                                response.putString("deeplink", attributionData.getDeeplink());
                            }
                            
                            // Convert campaign data to a WritableMap
                            if (attributionData.getCampaignData() != null) {
                                WritableMap campaignDataMap = Arguments.createMap();
                                campaignDataMap.putString("id", attributionData.getCampaignData().getId());
                                campaignDataMap.putString("name", attributionData.getCampaignData().getName());
                                
                                if (attributionData.getCampaignData().getAdNetwork() != null) {
                                    campaignDataMap.putString("adNetwork", attributionData.getCampaignData().getAdNetwork());
                                }
                                
                                campaignDataMap.putString("type", attributionData.getCampaignData().getType());
                                campaignDataMap.putString("installedAt", attributionData.getCampaignData().getInstalledAt());
                                
                                if (attributionData.getCampaignData().getStoreClickAt() != null) {
                                    campaignDataMap.putString("storeClickAt", attributionData.getCampaignData().getStoreClickAt());
                                }
                                
                                campaignDataMap.putString("groupName", attributionData.getCampaignData().getGroupName());
                                campaignDataMap.putString("assetName", attributionData.getCampaignData().getAssetName());
                                campaignDataMap.putString("assetGroupName", attributionData.getCampaignData().getAssetGroupName());
                                
                                response.putMap("campaignData", campaignDataMap);
                                
                            }
                        }
                        
                        response.putString("status", "success");
                        response.putString("message", "Attribution data retrieved successfully");
                        
                        promise.resolve(response);
                    } catch (Exception e) {
                        promise.reject("ATTRIBUTION_DATA_ERROR", "Error processing attribution data: " + e.getMessage(), e);
                    }
                    return Unit.INSTANCE;
                },
                exception -> {
                    promise.reject("ATTRIBUTION_DATA_ERROR", "Failed to get attribution data: " + exception.getMessage(), exception);
                    return Unit.INSTANCE;
                }
            );
        } catch (Exception e) {
            promise.reject("ATTRIBUTION_DATA_ERROR", "Failed to get attribution data: " + e.getMessage(), e);
        }
    }
    
    @ReactMethod
    public void setAdditionalData(ReadableMap integrationDataMap, Promise promise) {
        try {
            if (integrationDataMap == null || integrationDataMap.toHashMap().isEmpty()) {
                promise.reject("ADDITIONAL_DATA_ERROR", "Integration data is required");
                return;
            }
            
            // Create IntegrationData instance
            IntegrationData integrationData = new IntegrationData();
            
            // Extract clevertapId if present
            if (integrationDataMap.hasKey("clevertapId")) {
                integrationData.setClevertapId(integrationDataMap.getString("clevertapId"));
            }
            
            // Add other integration data fields here as they are added to the SDK
            
            // Call native SDK setAdditionalDataFromJava method
            linkrunnerSDK.setAdditionalDataFromJava(
                integrationData,
                () -> {
                    WritableMap response = Arguments.createMap();
                    response.putString("status", "success");
                    response.putString("message", "Additional data set successfully");
                    promise.resolve(response);
                    return Unit.INSTANCE;
                },
                exception -> {
                    promise.reject("ADDITIONAL_DATA_ERROR", "Failed to set additional data: " + exception.getMessage(), exception);
                    return Unit.INSTANCE;
                }
            );
        } catch (Exception e) {
            promise.reject("ADDITIONAL_DATA_ERROR", "Failed to set additional data: " + e.getMessage(), e);
        }
    }
}

File: android/src/main/java/io/linkrunner/LinkrunnerModule.kt
package io.linkrunner

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import io.linkrunner.sdk.BuildConfig
import io.linkrunner.sdk.LinkRunner
import io.linkrunner.sdk.models.request.UserDataRequest
import io.linkrunner.sdk.models.IntegrationData
import io.linkrunner.utils.MapUtils
import io.linkrunner.utils.ModelConverter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class LinkrunnerModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    
    private val TAG = "LinkrunnerModule"
    private val linkrunnerSDK = LinkRunner.getInstance()
    private val moduleScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    override fun getName(): String {
        return "LinkrunnerSDK"
    }

    @ReactMethod
    fun init(token: String, options: ReadableMap?, promise: Promise) {
        try {
            // Extract optional parameters
            val secretKey = options?.getString("secretKey")
            val keyId = options?.getString("keyId")
            val debug = options?.getBoolean("debug") ?: false

            val packageVersion = options?.getString("packageVersion") ?: "2.4.1" // React Native package version

            if (token.isEmpty()) {
                promise.reject("INIT_ERROR", "Token is required")
                return
            }

            moduleScope.launch {
                try {
                    // Configure SDK with client platform and version prior to init
                    try {
                        val platform = "REACT_NATIVE"
                        val packageVersion = packageVersion // React Native package version
                        LinkRunner.configureSDK(platform, packageVersion)
                    } catch (e: Exception) {
                        Log.w(TAG, "configureSDK failed: ${e.message}")
                    }
                    
                    val result = linkrunnerSDK.init(
                        context = reactContext,
                        token = token,
                        secretKey = secretKey,
                        keyId = keyId,
                        debug = debug
                    )
                    
                    withContext(Dispatchers.Main) {
                        if (result.isSuccess) {
                            val response = Arguments.createMap()
                            response.putString("status", "success")
                            response.putString("message", "Linkrunner SDK initialized successfully")
                            promise.resolve(response)
                        } else {
                            promise.reject(
                                "INIT_ERROR", 
                                "Failed to initialize Linkrunner: ${result.exceptionOrNull()?.message}",
                                result.exceptionOrNull()
                            )
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        promise.reject("INIT_ERROR", "Exception calling init: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            promise.reject("INIT_ERROR", "Failed to initialize Linkrunner: ${e.message}", e)
        }
    }

    @ReactMethod
    fun signup(userData: ReadableMap, data: ReadableMap?, promise: Promise) {
        try {
            // Convert ReadableMap to native Map for userData
            val userDataMap = MapUtils.readableMapToMap(userData)

            // Convert ReadableMap to native Map for additionalData
            val additionalData = if (data != null && !data.toHashMap().isEmpty()) {
                MapUtils.readableMapToMap(data)
            } else {
                null
            }

            moduleScope.launch {
                try {
                    val result = linkrunnerSDK.signup(
                        userData = ModelConverter.toUserDataRequest(userDataMap),
                        additionalData = additionalData
                    )
                    
                    withContext(Dispatchers.Main) {
                        if (result.isSuccess) {
                            val response = Arguments.createMap()
                            response.putString("status", "success")
                            response.putString("message", "User signed up successfully")
                            promise.resolve(response)
                        } else {
                            promise.reject(
                                "SIGNUP_ERROR", 
                                "Failed to signup user: ${result.exceptionOrNull()?.message}",
                                result.exceptionOrNull()
                            )
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        promise.reject("SIGNUP_ERROR", "Exception during signup: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            promise.reject("SIGNUP_ERROR", "Failed to signup user: ${e.message}", e)
        }
    }

    @ReactMethod
    fun setUserData(userData: ReadableMap, promise: Promise) {
        try {
            // Convert ReadableMap to native Map for userData
            val userDataMap = MapUtils.readableMapToMap(userData)

            moduleScope.launch {
                try {
                    val result = linkrunnerSDK.setUserData(ModelConverter.toUserDataRequest(userDataMap))
                    
                    withContext(Dispatchers.Main) {
                        if (result.isSuccess) {
                            val response = Arguments.createMap()
                            response.putString("status", "success")
                            response.putString("message", "User data set successfully")
                            promise.resolve(response)
                        } else {
                            promise.reject(
                                "SET_USER_DATA_ERROR", 
                                "Failed to set user data: ${result.exceptionOrNull()?.message}",
                                result.exceptionOrNull()
                            )
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        promise.reject("SET_USER_DATA_ERROR", "Exception setting user data: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            promise.reject("SET_USER_DATA_ERROR", "Failed to set user data: ${e.message}", e)
        }
    }

    @ReactMethod
    fun trackEvent(eventName: String, eventData: ReadableMap?, promise: Promise) {
        if (eventName.isEmpty()) {
            promise.reject("TRACK_EVENT_ERROR", "Event name is required")
            return
        }

        try {
            val eventDataMap = if (eventData != null) {
                MapUtils.readableMapToMap(eventData)
            } else {
                emptyMap()
            }

            moduleScope.launch {
                try {
                    val result = linkrunnerSDK.trackEvent(eventName, eventDataMap)
                    
                    withContext(Dispatchers.Main) {
                        if (result.isSuccess) {
                            val response = Arguments.createMap()
                            response.putString("status", "success")
                            response.putString("message", "Event tracked successfully")
                            promise.resolve(response)
                        } else {
                            promise.reject(
                                "TRACK_EVENT_ERROR", 
                                "Failed to track event: ${result.exceptionOrNull()?.message}",
                                result.exceptionOrNull()
                            )
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        promise.reject("TRACK_EVENT_ERROR", "Exception tracking event: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            promise.reject("TRACK_EVENT_ERROR", "Failed to track event: ${e.message}", e)
        }
    }

    @ReactMethod
    fun capturePayment(paymentData: ReadableMap, promise: Promise) {
        if (paymentData == null) {
            promise.reject("PAYMENT_ERROR", "Payment data is required")
            return
        }

        try {
            val paymentDataMap = MapUtils.readableMapToMap(paymentData)
            
            moduleScope.launch {
                try {
                    val result = linkrunnerSDK.capturePayment(ModelConverter.toCapturePaymentRequest(paymentDataMap))
                    
                    withContext(Dispatchers.Main) {
                        if (result.isSuccess) {
                            val response = Arguments.createMap()
                            response.putString("status", "success")
                            response.putString("message", "Payment captured successfully")
                            promise.resolve(response)
                        } else {
                            promise.reject(
                                "PAYMENT_ERROR", 
                                "Failed to capture payment: ${result.exceptionOrNull()?.message}",
                                result.exceptionOrNull()
                            )
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        promise.reject("PAYMENT_ERROR", "Exception capturing payment: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            promise.reject("PAYMENT_ERROR", "Failed to capture payment: ${e.message}", e)
        }
    }

    @ReactMethod
    fun removePayment(paymentData: ReadableMap, promise: Promise) {
        if (paymentData == null) {
            promise.reject("REMOVE_PAYMENT_ERROR", "Payment data is required")
            return
        }

        try {
            val paymentDataMap = MapUtils.readableMapToMap(paymentData)
            
            moduleScope.launch {
                try {
                    val result = linkrunnerSDK.removePayment(ModelConverter.toRemovePaymentRequest(paymentDataMap))
                    
                    withContext(Dispatchers.Main) {
                        if (result.isSuccess) {
                            val response = Arguments.createMap()
                            response.putString("status", "success")
                            response.putString("message", "Payment removed successfully")
                            promise.resolve(response)
                        } else {
                            promise.reject(
                                "REMOVE_PAYMENT_ERROR", 
                                "Failed to remove payment: ${result.exceptionOrNull()?.message}",
                                result.exceptionOrNull()
                            )
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        promise.reject("REMOVE_PAYMENT_ERROR", "Exception removing payment: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            promise.reject("REMOVE_PAYMENT_ERROR", "Failed to remove payment: ${e.message}", e)
        }
    }

    @ReactMethod
    fun getAttributionData(promise: Promise) {
        moduleScope.launch {
            try {
                val result = linkrunnerSDK.getAttributionData()
                
                withContext(Dispatchers.Main) {
                    if (result.isSuccess) {
                        val attributionData = result.getOrNull()
                        
                        // Convert the attribution data to a WritableMap
                        val response = Arguments.createMap()
                        
                        if (attributionData != null) {
                            // Add the deeplink if it exists
                            attributionData.deeplink?.let { deeplink ->
                                response.putString("deeplink", deeplink)
                            }
                            
                            // Convert campaign data to a WritableMap if it exists
                            attributionData.campaignData?.let { campaignData ->
                                val campaignDataMap = Arguments.createMap()
                                campaignDataMap.putString("id", campaignData.id)
                                campaignDataMap.putString("name", campaignData.name)
                                
                                campaignData.adNetwork?.let { adNetwork ->
                                    campaignDataMap.putString("adNetwork", adNetwork)
                                }
                                
                                campaignDataMap.putString("type", campaignData.type)
                                campaignDataMap.putString("installedAt", campaignData.installedAt)
                                
                                campaignData.storeClickAt?.let { storeClickAt ->
                                    campaignDataMap.putString("storeClickAt", storeClickAt)
                                }
                                
                                campaignDataMap.putString("groupName", campaignData.groupName)
                                campaignDataMap.putString("assetName", campaignData.assetName)
                                campaignDataMap.putString("assetGroupName", campaignData.assetGroupName)
                                
                                response.putMap("campaignData", campaignDataMap)
                            }
                        }
                        
                        response.putString("status", "success")
                        response.putString("message", "Attribution data retrieved successfully")
                        
                        promise.resolve(response)
                    } else {
                        promise.reject(
                            "ATTRIBUTION_DATA_ERROR", 
                            "Failed to get attribution data: ${result.exceptionOrNull()?.message}",
                            result.exceptionOrNull()
                        )
                    }
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    promise.reject("ATTRIBUTION_DATA_ERROR", "Exception getting attribution data: ${e.message}", e)
                }
            }
        }
    }

    @ReactMethod
    fun setAdditionalData(integrationDataMap: ReadableMap?, promise: Promise) {
        if (integrationDataMap == null || integrationDataMap.toHashMap().isEmpty()) {
            promise.reject("ADDITIONAL_DATA_ERROR", "Integration data is required")
            return
        }
        
        try {
            val additionalData = MapUtils.readableMapToMap(integrationDataMap)
            
            moduleScope.launch {
                try {
                    // Convert Map to IntegrationData
                    val integrationData = ModelConverter.toIntegrationData(additionalData)
                    val result = linkrunnerSDK.setAdditionalData(integrationData)
                    
                    withContext(Dispatchers.Main) {
                        if (result.isSuccess) {
                            val response = Arguments.createMap()
                            response.putString("status", "success")
                            response.putString("message", "Additional data set successfully")
                            promise.resolve(response)
                        } else {
                            promise.reject(
                                "ADDITIONAL_DATA_ERROR", 
                                "Failed to set additional data: ${result.exceptionOrNull()?.message}",
                                result.exceptionOrNull()
                            )
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        promise.reject("ADDITIONAL_DATA_ERROR", "Exception setting additional data: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            promise.reject("ADDITIONAL_DATA_ERROR", "Failed to set additional data: ${e.message}", e)
        }
    }

    @ReactMethod
    fun enablePIIHashing(enabled: Boolean) {
        try {
            // Call the SDK's enablePIIHashing method
            linkrunnerSDK.enablePIIHashing(enabled)
            
            if (BuildConfig.DEBUG) {
                Log.d(TAG, "Linkrunner: PII hashing ${if (enabled) "enabled" else "disabled"}")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to ${if (enabled) "enable" else "disable"} PII hashing", e)
        }
    }

    @ReactMethod
    fun setPushToken(pushToken: String, promise: Promise) {
        if (pushToken.isBlank()) {
            promise.reject("SET_PUSH_TOKEN_ERROR", "Push token cannot be empty")
            return
        }

        try {
            moduleScope.launch {
                try {
                    val result = linkrunnerSDK.setPushToken(pushToken)
                    
                    withContext(Dispatchers.Main) {
                        if (result.isSuccess) {
                            val response = Arguments.createMap()
                            response.putString("status", "success")
                            response.putString("message", "Push token set successfully")
                            promise.resolve(response)
                        } else {
                            promise.reject(
                                "SET_PUSH_TOKEN_ERROR", 
                                "Failed to set push token: ${result.exceptionOrNull()?.message}",
                                result.exceptionOrNull()
                            )
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) {
                        promise.reject("SET_PUSH_TOKEN_ERROR", "Exception setting push token: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            promise.reject("SET_PUSH_TOKEN_ERROR", "Failed to set push token: ${e.message}", e)
        }
    }
}
